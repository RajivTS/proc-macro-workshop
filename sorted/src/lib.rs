use quote::ToTokens;
use std::cmp::Ordering;

#[proc_macro_attribute]
pub fn sorted(
    args: proc_macro::TokenStream,
    input: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    let _ = args;
    let _ = input;
    let input = syn::parse_macro_input!(input as syn::Item);
    validate_enum_type(input)
        .and_then(|v| validate_enum_ordering(v))
        .map(|v| v.into_token_stream())
        .unwrap_or_else(syn::Error::into_compile_error)
        .into()
}

fn validate_enum_ordering(enum_item: syn::ItemEnum) -> syn::Result<syn::ItemEnum> {
    let mut out_of_place = None;
    let mut prev_variant = None;
    for variant in enum_item.variants.iter() {
        match prev_variant {
            None => prev_variant = Some(variant),
            Some(previous_variant) => {
                if let Ordering::Less = variant.ident.cmp(&previous_variant.ident) {
                    out_of_place = Some(variant);
                    break;
                }
                prev_variant = Some(variant)
            }
        }
    }
    match out_of_place {
        None => Ok(enum_item),
        Some(out_of_place) => {
            let swap_with = enum_item
                .variants
                .iter()
                .find(|&variant| out_of_place.ident.cmp(&variant.ident) == Ordering::Less)
                .unwrap_or(out_of_place);
            Err(syn::Error::new_spanned(
                &out_of_place.ident,
                format!(
                    "{} should sort before {}",
                    out_of_place.ident, swap_with.ident
                ),
            ))
        }
    }
}

fn validate_enum_type(item: syn::Item) -> syn::Result<syn::ItemEnum> {
    let attrs = match item {
        syn::Item::Const(const_item) => const_item.attrs,
        syn::Item::Enum(enum_item) => return Ok(enum_item),
        syn::Item::ExternCrate(crate_item) => crate_item.attrs,
        syn::Item::Fn(fn_item) => fn_item.attrs,
        syn::Item::ForeignMod(foreign_item) => foreign_item.attrs,
        syn::Item::Impl(impl_item) => impl_item.attrs,
        syn::Item::Macro(macro_item) => macro_item.attrs,
        syn::Item::Macro2(macro2_item) => macro2_item.attrs,
        syn::Item::Mod(mod_item) => mod_item.attrs,
        syn::Item::Static(static_item) => static_item.attrs,
        syn::Item::Struct(struct_item) => struct_item.attrs,
        syn::Item::Trait(trait_item) => trait_item.attrs,
        syn::Item::TraitAlias(trait_alias_item) => trait_alias_item.attrs,
        syn::Item::Type(type_item) => type_item.attrs,
        syn::Item::Union(union_item) => union_item.attrs,
        syn::Item::Use(use_item) => use_item.attrs,
        _ => {
            return Err(syn::Error::new_spanned(
                item,
                "expected enum or match expression",
            ))
        }
    };
    return Err(syn::Error::new_spanned(
        attrs.first(),
        "expected enum or match expression",
    ));
}
