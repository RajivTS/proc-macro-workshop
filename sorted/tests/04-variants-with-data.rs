// This test is similar to the previous where want to ensure that the macro
// correctly generates an error when the input enum is out of order, but this
// time it is using an enum that also has data associated with each variant.

use sorted::sorted;

#[sorted]
pub enum Error {
    Fmt(fmt::Error),
    Io(io::Error),
    Utf8(Utf8Error),
    Var(VarError),
    Dyn(Box<dyn StdError>),
}

fn main() {}
