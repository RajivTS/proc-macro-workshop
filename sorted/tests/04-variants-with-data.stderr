error: Dyn should sort before Fmt
  --> tests/04-variants-with-data.rs:13:5
   |
13 |     Dyn(Box<dyn StdError>),
   |     ^^^
